File: app/src/main/java/ge/dev/baqari/myfit/component/StepService.java
package ge.dev.baqari.myfit.component;

import android.app.AlarmManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Intent;
import android.os.Build;
import android.os.IBinder;

import androidx.core.app.NotificationCompat;

import org.greenrobot.eventbus.EventBus;

import ge.dev.baqari.myfit.api.StepThread;
import ge.dev.baqari.myfit.utils.LocalKeys;
import ge.dev.baqari.myfit.utils.NotificationExtentionKt;
import ge.dev.baqari.myfit.utils.SharedPrefsKt;

public class StepService extends Service {

    private StepThread stepThread;
    private boolean startedForeground;

    public static final String RESTART_ACTION = "ge.dev.baqari.myfit.RESTART";
    public static final String STOP_REMOTELY = "ge.dev.baqari.myfit.REMOTELY";
    public static final String START = "ge.dev.baqari.myfit.START";

    public boolean getNotificationEnabled() {
        return SharedPrefsKt.get().getBoolean(LocalKeys.INSTANCE.getNotificationEnabledKey(), true);
    }

    public void setNotificationEnabled(boolean enabled) {
        SharedPrefsKt.get().edit().putBoolean(LocalKeys.INSTANCE.getNotificationEnabledKey(), enabled).apply();
    }

    @Override
    public void onCreate() {
        super.onCreate();
        stepThread = new StepThread(this);
        if (Build.VERSION.SDK_INT >= 26 && getNotificationEnabled()) {
            startPushForeground(null);
        }
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if (intent != null) {
            if (stepThread == null)
                stepThread = new StepThread(this);
            stepThread.run();

            if (intent.getAction() != null) {
                if (intent.getAction().equals(STOP_REMOTELY)) {
                    EventBus.getDefault().post(false);
                    setNotificationEnabled(false);
                    stopForeground(true);
                } else {
                    if (intent.getAction().equals(RESTART_ACTION) || intent.getAction().equals(START))
                        startedForeground = false;

                    if (!startedForeground) {
                        stepThread.setOnStep(new StepThread.OnStepUpdateListener() {
                            @Override
                            public void onStep(long step) {
                                if (getNotificationEnabled()) {
                                    startPushForeground(step);
                                    startedForeground = true;
                                }
                                fireWakeLock();
                            }
                        });
                        stepThread.invokeOnStep();
                        if (getNotificationEnabled()) {
                            startPushForeground(null);
                            startedForeground = true;
                        }
                        fireWakeLock();
                    }
                }
            }
        }
        return START_STICKY;
    }

    private void startPushForeground(Long stepCount) {
        NotificationCompat.Builder notificationBuilder = NotificationExtentionKt.startStepPushForeground(this, stepCount);
        startForeground(1024, notificationBuilder.build());
        EventBus.getDefault().post(true);
    }

    private void fireWakeLock() {
        //if (getNotificationEnabled()) {
        AlarmManager alarmManager = (AlarmManager) getSystemService(ALARM_SERVICE);
        if (alarmManager != null) {
            Intent intent = new Intent(this, BootBroadcastReceiver.class);
            intent.setAction(START);
            alarmManager.setInexactRepeating(AlarmManager.RTC_WAKEUP, System.currentTimeMillis(), 1000,
                    PendingIntent.getBroadcast(this, 0, intent, 0));
        }
        //}
    }


    @Override
    public void onTaskRemoved(Intent removedIntent) {
        super.onTaskRemoved(removedIntent);
        stopForeground(true);
        stepThread.threadStop();
        stepThread = null;
        //if (getNotificationEnabled()) {
        Intent intent = new Intent(this, BootBroadcastReceiver.class);
        intent.setAction(START);
        sendBroadcast(intent);
        //}
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        stopForeground(true);
        stepThread.threadStop();
        stepThread = null;
        //if (getNotificationEnabled()) {
        Intent intent = new Intent(this, BootBroadcastReceiver.class);
        intent.setAction(START);
        sendBroadcast(intent);
        //}
    }

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }
}

File: app/src/main/java/com/mygpi/mygpimobilefitness/component/StepService.kt
package com.mygpi.mygpimobilefitness.component

import android.annotation.SuppressLint
import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.os.Build
import android.os.IBinder
import android.os.PowerManager
import android.widget.RemoteViews
import androidx.annotation.RequiresApi
import androidx.core.app.NotificationCompat
import androidx.core.app.TaskStackBuilder

import com.mygpi.mygpimobilefitness.R
import com.mygpi.mygpimobilefitness.api.BaseCalculator
import com.mygpi.mygpimobilefitness.api.StepThread
import com.mygpi.mygpimobilefitness.round
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode

@SuppressLint("Registered")
class StepService : Service() {
    private var thread: StepThread? = null
    private var mWakeLock: PowerManager.WakeLock? = null
    private var startedForeground: Boolean = false
    private var kilometersCount = 0.0

    override fun onCreate() {
        super.onCreate()
        thread = StepThread(this)
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        if (thread?.state == Thread.State.NEW)
            thread?.start()

        if (!startedForeground) {
            mWakeLock(this)
            startedForeground = true
            thread?.onStep = {
                startForeground(it)
            }
            thread?.invokeOnStep()
        }

        return START_STICKY
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    fun startForeground(stepCount: Long) {
        var channelId = ""
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val notificationManager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            val stepCounterChannel = NotificationChannel("step_counter_channel_id",
                    "step_counter_channel_id", NotificationManager.IMPORTANCE_DEFAULT)
            stepCounterChannel.enableVibration(true)
            stepCounterChannel.lightColor = Color.RED
            stepCounterChannel.lockscreenVisibility = Notification.VISIBILITY_PUBLIC
            channelId = stepCounterChannel.id
            notificationManager.createNotificationChannel(stepCounterChannel)
        }

        val notificationLayout = RemoteViews(packageName, R.layout.notification_layout)
        val notificationLayoutExpanded = RemoteViews(packageName, R.layout.notification_layout)

        kilometersCount += BaseCalculator.calculateKilometers(stepCount)
        notificationLayout.setTextViewText(R.id.notificationStepCount, "${getString(R.string.steps_)} $stepCount")
        notificationLayout.setTextViewText(R.id.notificationKmCount, "${getString(R.string.km_s)} ${kilometersCount.round(2)}")

        notificationLayoutExpanded.setTextViewText(R.id.notificationStepCount, "${getString(R.string.steps_)} $stepCount")
        notificationLayoutExpanded.setTextViewText(R.id.notificationKmCount, "${getString(R.string.km_s)} ${kilometersCount.round(2)}")

        val mBuilder = NotificationCompat.Builder(this, channelId)
                .setStyle(NotificationCompat.DecoratedCustomViewStyle())
                .setCustomContentView(notificationLayout)
                .setCustomBigContentView(notificationLayoutExpanded)
                .setSmallIcon(R.mipmap.ic_launcher)

        val notificationIntent = Intent(this, MainActivity::class.java)
        val stackBuilder = TaskStackBuilder.create(this)
        stackBuilder.addParentStack(MainActivity::class.java)
        stackBuilder.addNextIntent(notificationIntent)
        val resultPendingIntent = stackBuilder.getPendingIntent(0, PendingIntent.FLAG_UPDATE_CURRENT)
        mBuilder.setContentIntent(resultPendingIntent)

        startForeground(1024, mBuilder.build())
    }

    override fun onTaskRemoved(rootIntent: Intent) {
        super.onTaskRemoved(rootIntent)
        if (mWakeLock != null) {
            if (mWakeLock?.isHeld == true)
                mWakeLock?.release()
            mWakeLock = null
        }

        stopForeground(true)
        thread?.threadStop()
        val intent = Intent(this, BootBroadcastReceiver::class.java).apply { action = START }
        sendBroadcast(intent)
    }

    override fun onDestroy() {
        if (mWakeLock != null) {
            if (mWakeLock?.isHeld == true)
                mWakeLock?.release()
            mWakeLock = null
        }

        stopForeground(true)
        thread?.threadStop()
        val intent = Intent(this, BootBroadcastReceiver::class.java).apply { action = START }
        sendBroadcast(intent)
        super.onDestroy()
    }


    override fun onBind(intent: Intent): IBinder? {
        return null
    }

    @Synchronized
    private fun mWakeLock(context: Context) {
        if (mWakeLock != null) {
            if (mWakeLock?.isHeld == true)
                mWakeLock?.release()
            mWakeLock = null
        }

        val mgr = context.getSystemService(Context.POWER_SERVICE) as PowerManager
        mWakeLock = mgr.newWakeLock(PowerManager.PARTIAL_WAKE_LOCK, StepService::class.java.name)
        mWakeLock?.setReferenceCounted(true)
        mWakeLock?.acquire(1000L)
    }

    companion object {
        val START = "com.mygpi.mygpimobilefitness.start"
    }
}